src/dotnet/ToolPackage/ToolPackageInstance.cs

```diff
@@ -28,6 +32,58 @@ public IReadOnlyList<CommandSettings> Commands
             }
         }
 
+        public IReadOnlyList<FilePath> PackagedShims
```

Comment:
I think this is too complicated (and expensive) for a property.  I think this should either be a `Lazy` initialized field or a method called `GetPackagedSims` to indicate it is performing work (I favor the first solution since this list shouldn't change).

Reply:
i genuinely confused with the line of Get vs property. In my mind most of the query methods in ToolPackageInstance do disk IO. For me that's the bar. However, do you think we should all convert it to method? Or because of most of them are Lazy so that is fine?

Reply:
It's not that it's inherently incorrect, I'm just of the opinion that developers tend to think of properties as "cheap" (return a field, for example) or "expensive at most-once" (lazy-init).  Having a property do work every time can be very surprising for users, especially for enumerable properties:

``` 
foreach (var shim in package.PackagedShims)
{
}

...

foreach (var shim in package.PackagedShims)
{
}
```

As someone looking at that code, I'd expect the two enumerables to be the same.  If I saw a method called `GetPackagedShims` or `EnumeratePackagedShims` instead, I'd expect each invocation to return a different enumeration.

Reply:
"expensive-at-most-once" is the right bar for properties. Anything that does a bunch of computation or IO whenever invoked should be a method.

It looks like a field access, so it should behave like a field access. There's nothing immediately wrong looking with the following. (Other than "style nit: use foreach, but use your imagination to think of a case where indexing is actually needed.)

``` C#
for (int i = 0, i < package.PackagedShims.Count; i++)
{
     var x = package.PackagedShims[i];
}
```

But it calls package.get_PackageShims() twice per loop iteration!

Reply:
Fixed, use Lazy